x/bank/client/cli/query.go

```diff
@@ -41,76 +43,108 @@ func GetBalancesCmd(clientCtx client.Context) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "balances [address]",
 		Short: "Query for account balances by address",
-		Args:  cobra.ExactArgs(1),
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query the total balance of an account or of a specific denomination.
+
+Example:
+  $ %s query %s balances [address]
+  $ %s query %s balances [address] --denom=[denom]
+`,
+				version.ClientName, types.ModuleName, version.ClientName, types.ModuleName,
+			),
+		),
+		Args: cobra.ExactArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			queryClient := types.NewQueryClient(clientCtx.Init())
+			clientCtx, err := client.ReadGetCommandFlags(clientCtx, cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			denom, err := cmd.Flags().GetString(FlagDenom)
+			if err != nil {
+				return err
+			}
+
+			queryClient := types.NewQueryClient(clientCtx)
 
 			addr, err := sdk.AccAddressFromBech32(args[0])
 			if err != nil {
 				return err
 			}
 
-			denom := viper.GetString(flagDenom)
 			pageReq := &query.PageRequest{}
 			if denom == "" {
 				params := types.NewQueryAllBalancesRequest(addr, pageReq)
+
 				res, err := queryClient.AllBalances(context.Background(), params)
 				if err != nil {
 					return err
 				}
-				return clientCtx.PrintOutput(res.Balances)
 
+				return clientCtx.PrintOutput(res.Balances)
 			}
 
 			params := types.NewQueryBalanceRequest(addr, denom)
 			res, err := queryClient.Balance(context.Background(), params)
 			if err != nil {
 				return err
 			}
+
 			return clientCtx.PrintOutput(res.Balance)
 		},
 	}
 
-	cmd.Flags().String(flagDenom, "", "The specific balance denomination to query for")
+	cmd.Flags().String(FlagDenom, "", "The specific balance denomination to query for")
 	return flags.GetCommands(cmd)[0]
 }
 
 func GetCmdQueryTotalSupply(clientCtx client.Context) *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "total [denom]",
-		Args:  cobra.MaximumNArgs(1),
+		Use:   "total",
 		Short: "Query the total supply of coins of the chain",
 		Long: strings.TrimSpace(
-			fmt.Sprintf(`Query total supply of coins that are held by accounts in the
-			chain.
+			fmt.Sprintf(`Query total supply of coins that are held by accounts in the chain.
 
 Example:
-$ %s query %s total
+  $ %s query %s total
 
 To query for the total supply of a specific coin denomination use:
-$ %s query %s total stake
+  $ %s query %s total --denom=[denom]
 `,
 				version.ClientName, types.ModuleName, version.ClientName, types.ModuleName,
 			),
 		),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			queryClient := types.NewQueryClient(clientCtx.Init())
+			clientCtx, err := client.ReadGetCommandFlags(clientCtx, cmd.Flags())
```

Comment:
I think this is cleaner now, yet still simple. What are you suggesting to improve on top of this @aaronc?

Reply:
I like this approach